bin/shared.sh

```diff
@@ -0,0 +1,187 @@
+# `shared.bat` script in the same directory to ensure that Flutter & Dart continue
+# to work across all platforms!
+#
+# -------------------------------------------------------------------------- #
+
+set -e
+
+unset CDPATH
+
+function _rmlock () {
+  [ -n "$FLUTTER_UPGRADE_LOCK" ] && rm -f "$FLUTTER_UPGRADE_LOCK"
+}
+
+function retry_upgrade {
+  local total_tries="10"
+  local remaining_tries=$((total_tries - 1))
+  while [[ "$remaining_tries" -gt 0 ]]; do
+    (cd "$FLUTTER_TOOLS_DIR" && "$PUB" upgrade "$VERBOSITY" --no-precompile) && break
+    echo "Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... ($remaining_tries tries left)"
+    remaining_tries=$((remaining_tries - 1))
+    sleep 5
+  done
+
+  if [[ "$remaining_tries" == 0 ]]; then
+    echo "Command 'pub upgrade' still failed after $total_tries tries, giving up."
+    return 1
+  fi
+  return 0
+}
+
+function upgrade_flutter () {
+  mkdir -p "$FLUTTER_ROOT/bin/cache"
+
+  # This function is executed with a redirect that pipes the source of
+  # this script into file descriptor 3.
+  #
+  # To ensure that we don't simultaneously update Dart in multiple
+  # parallel instances, we try to obtain an exclusive lock on this
+  # file descriptor (and thus this script's source file) while we are
+  # updating Dart and compiling the script. To do this, we try to use
+  # the command line program "flock", which is available on many
+  # Unix-like platforms, in particular on most Linux distributions.
+  # You give it a file descriptor, and it locks the corresponding
+  # file, having inherited the file descriptor from the shell.
+  #
+  # Complicating matters, there are two major scenarios where this
+  # will not work.
+  #
+  # The first is if the platform doesn't have "flock", for example on Mac.
+  # There is not a direct equivalent, so on platforms that don't have flock,
+  # we fall back to using a lockfile and spinlock with "shlock".  This
+  # doesn't work as well over NFS as it relies on PIDs. Any platform
+  # without either of these tools has no locking at all. To determine if we
+  # have "flock" or "shlock" available, we abuse the "hash" shell built-in.
+  #
+  # The second complication is NFS. On NFS, to obtain an exclusive
+  # lock you need a file descriptor that is open for writing, because
+  # NFS implements exclusive locks by writing, or some such. Thus, we
+  # ignore errors from flock. We do so by using the '|| true' trick,
+  # since we are running in a 'set -e' environment wherein all errors
+  # are fatal, and by redirecting all output to /dev/null, since
+  # users will typically not care about errors from flock and are
+  # more likely to be confused by them than helped.
+  #
+  # For "flock", the lock is released when the file descriptor goes out of
+  # scope,  i.e. when this function returns.  The lock is released via
+  # a trap when using "shlock".
+  if hash flock 2>/dev/null; then
```

Comment:
What about using the mkdir method described in this FAQ, at least as a fallback when flock doesn't exist? It seems more reliable than shlock, and would probably work on platforms without either one: http://mywiki.wooledge.org/BashFAQ/045

It goes something like this:
```bash
 lockdir=/tmp/myscript.lock
 if mkdir "$lockdir"
 then
     echo >&2 "successfully acquired lock"

     # Remove lockdir when the script finishes, or when it receives a signal
     trap 'rm -rf "$lockdir"' 0    # remove directory when script finishes

     # Optionally create temporary files in this directory, because
     # they will be removed automatically:
     tmpfile=$lockdir/filelist

 else
     echo >&2 "cannot acquire lock, giving up on $lockdir"
     exit 0
 fi
```

Reply:
> Sure, that's fine. We should address them, though.

pinky swear :)

Reply:
Filed https://github.com/flutter/flutter/issues/57512 to track.

Reply:
There are some poorly implemented userland filesystems where the mkdir trick doesn't work, but those are edge cases on an edge case.  It's a pretty solid backup for flock.

Reply:
> There are some poorly implemented userland filesystems where the mkdir trick doesn't work, but those are edge cases on an edge case. It's a pretty solid backup for flock.

*sigh*, it seems like BASH scripting is all about choosing which edge cases you want to optimize for...

Reply:
There's another case that is an edge case and a problem: If I'm mounting the flutter dir over a file share, and I access it from both a Linux system and a Mac system simultaneously, then the locking won't hold, since one is using flock and one is using shlock (or mkdir). They'll happily overwrite each other then.

Reply:
This is true. Using the same repo from multiple host OSes is not supported anyway (the cached artifacts can't be shared), but we should detect this.

Reply:
BTW, I put up a draft PR to try and address some of this (although it doesn't address the two-hosts problem): https://github.com/flutter/flutter/pull/57590